src/HTTPure/RequestHeaders.purs

```diff
@@ -0,0 +1,48 @@
+module HTTPure.RequestHeaders
+  ( RequestHeaders(..)
+  , empty
+  , read
+  ) where
+
+import Prelude
+
+import Data.FoldableWithIndex (foldMapWithIndex)
+import Data.Newtype (class Newtype)
+import Data.String as String
+import Foreign.Object (Object, union)
+import Foreign.Object as Object
+import HTTPure.Lookup (class Lookup, (!!))
+import Node.HTTP (Request, requestHeaders)
+
+-- | The `RequestHeaders` type is just sugar for a `Object` of `Strings`
+-- | that represents the set of headers in an HTTP request.
+newtype RequestHeaders = RequestHeaders (Object String)
+
+derive instance newtypeRequestHeaders :: Newtype RequestHeaders _
+
+-- | Given a string. return a `Maybe` containing the value of the matching
+-- | request header, if there is any.
+instance lookupRequestHeaders :: Lookup RequestHeaders String String where
+  lookup (RequestHeaders headers') key = headers' !! (String.toLower key)
+
+-- | Allow a `RequestHeaders` to be represented as a string. This string is
+-- | formatted in HTTP headers format.
+instance showRequestHeaders :: Show RequestHeaders where
+  show (RequestHeaders headers') = foldMapWithIndex showField headers' <> "\n"
+    where
+    showField key value = key <> ": " <> value <> "\n"
+
+-- | Compare two `RequestHeaders` objects by comparing the underlying `Objects`.
+instance eqRequestHeaders :: Eq RequestHeaders where
+  eq (RequestHeaders a) (RequestHeaders b) = eq a b
+
+-- | Allow one `RequestHeaders` objects to be appended to another.
+instance semigroupRequestHeaders :: Semigroup RequestHeaders where
+  append (RequestHeaders a) (RequestHeaders b) = RequestHeaders $ union b a
+
+-- | Get the request headers out of a HTTP `Request` object.
+read :: Request -> RequestHeaders
+read = requestHeaders >>> RequestHeaders
```

Comment:
Seems like we're missing a way to access [array-like headers](https://nodejs.org/api/http.html#http_message_headers) like `set-cookie` here. Should we also have a `read'` that returns maybe a `RequestHeaders' (Object (Array String))`?

Reply:
The problem here is that we just can't know which headers are array-like.

Reply:
We can with FFI. That of course requires us to assume a JS backend, but we're already using Node bindings e.g. for `setHeaders`, so I think we've already made that assumption.

Reply:
We could have a function that returns `Either String (Array String)` for the headers, I guess, is that what you mean?

Reply:
Yeah, we have `read :: Request -> Object String`, we could also have `read' :: Request -> Object (Array String)`. Or we could have a more descriptive name like `readArray :: Request -> Object (Array String)`.

Reply:
It looks like if we go down this path, we'd be using node's `headersDistinct` property on the `Request` object, which returns an `Object (NonEmptyArray String)`, see the [node docs](https://nodejs.org/api/http.html#messageheadersdistinct). If we do that, then `RequestHeaders` will have the same shape as `ResponseHeaders` again, but we'll have solved the issue described in #157.

I'm considering closing this PR and implementing a new solution where we just change the type of `HTTPure.Headers.Headers` and don't create separate types for request and response headers.

Reply:
Haha yeah after 4975d1017413dd9ccfa4d171da009df2ca73f9b6, I started wondering why we need the distinction as well.

Reply:
@cprussin, you and @paluh [spoke offline](https://github.com/citizennet/purescript-httpure/issues/157#issuecomment-651394137) when you came up with the different header types solution, do you have any additional insight to offer here?

Reply:
@davezuch Actually, after thinking a bit, I believe we should not provide a way of reading duplicate request headers. Duplicate response headers are used by browsers, like with `Set-Cookie`, `Cache-Control`, `Access-Control-*`, etc. But I couldn't find examples of duplicate request headers online, so this would be a very application-specific thing; and for that there are alternatives, like parsing.

The thing is that changing the type of `RequestHeaders` to use `NonEmptyArray String` would cause a lot of breaking changes and would degrade the usability of the library a bit, as the users would essentially always have to apply `Data.Array.NonEmpty.head` to the result of the `Lookup` functions.

I'm considering just sticking to what we have here, but changing `RequestHeaders` to be a `Map CaseInsensitiveString String` instead of `Object String`; what do you think?

Reply:
@arthurxavierx I had in mind an API that supports both `String` and `NonEmptyArray String` headers, e.g.:

```purs
module HTTPure.Headers
  ( ArrayHeaders(..)
  , StringHeaders(..)
  , getArrayHeader
  , getArrayHeaders
  , getStringHeader
  , getStringHeaders
   ...
  ) where

type ArrayHeaders = Map CaseInsensitiveString (NonEmptyArray String)

type StringHeaders = Map CaseInsensitiveString String

getArrayHeader :: String -> Request -> Maybe (NonEmptyArray String)
getArrayHeader = ...

getArrayHeaders :: Request -> ArrayHeaders
getArrayHeadrs = ...

getStringHeader :: String -> Request -> Maybe String
getStringHeader = ...

getStringHeaders :: Request -> StringHeaders
getStringHeaders = ...
```

I was basing my suggestion off the [NodeJS API](https://nodejs.org/api/http.html#http_message_headers), which doesn't disambiguate between request and response headers, and instead cases on the header itself when deciding whether the value should be a string or array. Basically, instead of splitting `Headers` into `RequestHeaders` and `ResponseHeaders`, we'd split it into `StringHeaders` and `ArrayHeaders` (not necessarily those names).

It seems like a more flexible API to me, but I don't feel strongly about it. If you think sticking to what we have here is a better API, that works for me too.

Reply:
Well, either change is breaking, it's just breaking in different directions (
> Basically, instead of splitting `Headers` into `RequestHeaders` and `ResponseHeaders`, we'd split it into `StringHeaders` and `ArrayHeaders`

).

If we go with the direction of this PR, it probably wouldn't be worth it to adopt my suggestion later. It's more a one or the other choice IMO.

But like I said, I don't feel strongly here. It seems a more flexible API, but maybe there's no point in supporting array headers for requests.

Reply:
Thinking about it though, my suggestion might not require a breaking change. We just add a new `ArrayHeaders` type to `HTTPure.Headers`, and the supporting functions. Or even add a new module for `HTTPure.ArrayHeaders` (again, not set on the name, maybe `MultipleHeaders` or `HeadersCollection`, etc.), and then the current headers continue to work as-is.

Reply:
One major problem with this approach is that we'd also have to create `HTTPure.Request.RequestArray` and `HTTPure.Response.ResponseArray` variants, as currently `HTTPure.Request.Request` and `HTTPure.Request.Response` both have a `headers :: HTTPure.Headers.Headers` field.

Reply:
I think we would just extend them to contain both (I'm now thinking the name could be `MultiHeaders` instead of `ArrayHeaders`):

```diff
  type Request =
    { method :: Method
    , path :: Path
    , query :: Query
    , headers :: Headers
+   , multiHeaders :: MultiHeaders
    , httpVersion :: Version
    , url :: String
    }
```

```diff
  type Response =
    { status :: Status
    , headers :: Headers
+   , multiHeaders :: MultiHeaders
    , writeBody :: HTTP.Response -> Aff Unit
    }
```

Reply:
Hm I hadn't considered that option. Looks like it should work. Thanks!